cmake/modules/extensions.cmake

```diff
@@ -3332,6 +3332,49 @@ macro(zephyr_linker_memory_ifdef feature_toggle)
   endif()
 endmacro()
 
+# Usage:
+#   zephyr_linker_dts_section_from_path(PATH <path>)
+#
+# Zephyr linker devicetree memory section from path.
+#
+# This function specifies a memory section for the platform in use based on its
+# devicetree configuration.
+#
+# The memory will only be defined if the devicetree exists and has status okay.
+#
+# PATH <path>      : Devicetree node path.
+#
+function(zephyr_linker_dts_section_from_path)
+  set(single_args "PATH")
+  cmake_parse_arguments(DTS_SECTION "" "${single_args}" "" ${ARGN})
+
+  if(DTS_SECTION_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "zephyr_linker_dts_section_from_path(${ARGV0} ...) given unknown "
+	    "arguments: ${DTS_SECTION_UNPARSED_ARGUMENTS}"
+    )
+  endif()
+
+  if(NOT DEFINED DTS_SECTION_PATH)
+    return()
+  endif()
+
+  dt_node_exists(exists PATH ${DTS_SECTION_PATH})
+  if(NOT ${exists})
+    return()
+  endif()
+
+  dt_reg_addr(addr PATH ${DTS_SECTION_PATH})
+
+  dt_prop(name PATH ${DTS_SECTION_PATH} PROPERTY "zephyr,memory-region")
+  if (NOT DEFINED name)
+    # Fallback to the node path
+    set(name ${DTS_SECTION_PATH})
```

Comment:
shouldn't it be described above that if the node path points to a `zephyr,memory-region` property, then the name of the memory region will be used, else the node path is used as name.

Secondly, as this is a `path`, should we sanitize the `name` ?
And if not, then why not.

Reply:
uhm, we are back to the same discussion as https://github.com/zephyrproject-rtos/zephyr/pull/41055#issuecomment-1044372067.

We can be sure that the `zephyr,memory-region` property is present (because it is enforced by the YAML file) so the fallback on path is not actually valid. The problem is again when we need to sanitize the name.

When CMake is used we can use `zephyr_string()` as you suggested and that's easy enough, the problem is that I didn't find a good solution yet to sanitize the name or enforce `[a-zA-Z0-9]` for names.

Reply:
> uhm, we are back to the same discussion as https://github.com/zephyrproject-rtos/zephyr/pull/41055#issuecomment-1044372067.

I was thinking the same when I wrote the comment :laughing: 

Maybe for a start we should error out if `zephyr,memory-region` doesn't exists (or a warning stating that it was ignored).
Do we atm have a use-case where we want to allow the name be identical to path, when `zephyr,memory-region` is not defined ?

Afterall, if a user really wants to create a section with the name of a dts path , that is still possible by using:
```
dt_reg_addr(addr PATH ${my_dts_path}) 
zephyr_linker_section(NAME ${my_dts_path} ADDRESS ${addr} TYPE NOLOAD)
```

but user is then in full control and responsible that name is valid for linkers.

Reply:
> Maybe for a start we should error out if zephyr,memory-region doesn't exists (or a warning stating that it was ignored).

But the point is indeed that you **must** have the `zephyr,memory-region` property because that is marked as required by the DT binding, so you definitely always have the `name` available. The point is to make the `name` sanitized.

@mbolivar-nordic is there already any constraint on the DT strings (i.e. [a-zA-Z0-9])?

Reply:
> > Maybe for a start we should error out if zephyr,memory-region doesn't exists (or a warning stating that it was ignored).
> 
> But the point is indeed that you **must** have the `zephyr,memory-region` property

Not in you original implementation here: https://github.com/zephyrproject-rtos/zephyr/blob/eee0da7fc1d79ecc29657ec37c1c310ac56a1053/cmake/modules/extensions.cmake#L3368-L3372

cause that code doesn't verify that this property exists.
So users can call this with any node path, regardless of the properties they have.

Read the function description:
https://github.com/zephyrproject-rtos/zephyr/blob/5e79463ace625b649028e599e5ad75e52d3bc168/cmake/modules/extensions.cmake#L2698-L2700
https://github.com/zephyrproject-rtos/zephyr/blob/5e79463ace625b649028e599e5ad75e52d3bc168/cmake/modules/extensions.cmake#L2720-L2722

So if the property doesn't exists, then `name` is undefined.
And that even goes for the latest push.

Users can call
```
zephyr_linker_dts_section(PATH "/soc")
```
and no error will be returned by the function.

Reply:
Ok, I see what you mean. Tried to address that.